src/adc.rs

```diff
@@ -0,0 +1,562 @@
+//! # Work in Progress
+//! API for the ADC (Analog to Digital Converter)
+//!
+//! Note that the more specific your hardware selection is
+//! (e.g. stm32f303**xc** instead of just stm32f303)
+//! the more functionality is accessible
+//! (in this case: ADC3 and ADC4 additionaly to ADC1 and ADC2).
+//!
+//! # Examples
+//! For a simple, working example check `adc.rs` in the examples folder.
+use cortex_m::asm;
+use embedded_hal::adc::{Channel, OneShot};
+#[cfg(feature = "stm32f303")]
+use stm32f3::stm32f303::adc1_2::ccr::CKMODE_A;
+
+use crate::rcc::{Clocks, AHB};
+
+#[cfg(feature = "stm32f303")]
+const MAX_ADVREGEN_STARTUP_US: u32 = 10;
+
+#[cfg(feature = "stm32f303")]
+use crate::gpio::{gpioa, gpiob, gpioc, Analog};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::gpio::{gpiod, gpioe, gpiof};
+
+#[cfg(feature = "stm32f303")]
+use crate::stm32::{ADC1, ADC1_2, ADC2};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::stm32::{ADC3, ADC3_4, ADC4};
+
+/// ADC configuration
+///
+/// TODO: Remove `pub` from the register block once all functionalities are implemented.
+/// Leave it here until then as it allows easy access to the registers.
+pub struct Adc<ADC> {
+    pub rb: ADC,
+    clocks: Clocks,
+    ckmode: CKMODE,
+    operation_mode: Option<OperationMode>,
+}
+
+/// ADC sampling time
+///
+/// Each channel can be sampled with a different sample time.
+/// For Sampletime T_n the total conversion time (in ADC clock cycles) is
+/// 12.5 + (n + .5)
+///
+/// TODO: there are boundaries on how this can be set depending on the hardware.
+/// Check them and implement a sample time setting mechanism.
+pub enum SampleTime {
+    T_1,
+    T_2,
+    T_4,
+    T_7,
+    T_19,
+    T_61,
+    T_181,
+    T_601,
+}
+
+impl Default for SampleTime {
+    fn default() -> Self {
+        SampleTime::T_19
+    }
+}
+
+impl SampleTime {
+    /// Conversion to bits for SMP
+    fn bitcode(&self) -> u8 {
+        match self {
+            SampleTime::T_1 => 0b000,
+            SampleTime::T_2 => 0b001,
+            SampleTime::T_4 => 0b010,
+            SampleTime::T_7 => 0b011,
+            SampleTime::T_19 => 0b100,
+            SampleTime::T_61 => 0b101,
+            SampleTime::T_181 => 0b110,
+            SampleTime::T_601 => 0b111,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+/// ADC operation mode
+///
+/// TODO: Implement other modes (DMA, Differential,…)
+pub enum OperationMode {
+    OneShot,
+}
+
+#[derive(Clone, Copy, PartialEq)]
+/// ADC CKMODE
+///
+/// TODO: Add ASYNCHRONOUS mode
+pub enum CKMODE {
```

Comment:
Rust coding conventions suggest using CamelCase for type names. So something like `Ckmode` or `CkMode` would fit better